torch/distributed/optim/named_optimizer.py

```diff
@@ -0,0 +1,244 @@
+import logging
+import warnings
+
+from copy import deepcopy
+from typing import Any, Collection, Dict, List, Mapping, Union
+
+import torch
+from torch import optim
+from torch.distributed._shard.sharded_tensor import ShardedTensor
+
+
+__all__ = ["NamedOptimizer"]
+
+logger = logging.getLogger(__name__)
+
+
+class NamedOptimizer(optim.Optimizer):
+    """
+    NamedOptimizer takes a dict of parameters and exposes state_dict by parameter key.
+    We replace the original key (number) in an optim to the FQN string. User can
+    initialize the optim as they initialize a PyTorch optim, the only difference is
+    that they also need to passed in the FQN of each parameters.
+
+    Args:
+        named_parameters (Mapping[str, Union[torch.Tensor, ShardedTensor]]):
+            Parameters of the module and its FQN.
+        optimizer_class (optim.Optimizer):
+            the class of optimizer to instantiate.
+        param_groups (Collection[Mapping[str, Any]]):
+            param_groups to pass to optimizer if specified.
```

Comment:
Let us specify in this description that the keys in the inner `Mapping` are the FQNs? (In other words, from this description, it is not clear how to specify the `param_groups`.)

Reply:
Sure.